components/support/ffi/src/error.rs

```diff
@@ -0,0 +1,273 @@
+///
+/// ## Caveats
+///
+/// Note that the order of the fields is `code` (an i32) then `message` (a `*mut c_char`), getting
+/// this wrong on the other side of the FFI will cause memory corruption and crashes.
+///
+/// The fields are public largely for documentation purposes, but you should use
+/// [`ExternError::new_error`] or [`ExternError::success`] to create these.
+///
+/// ## Layout/fields
+///
+/// This struct's field are not `pub` (mostly so that we can soundly implement `Send`, but also so
+/// that we can verify rust users are constructing them appropriately), the fields, their types, and
+/// their order are *very much* a part of the public API of this type. Consumers on the other side
+/// of the FFI will need to know it's layout.
+///
+/// If this were a C struct, it would look like
+///
+/// ```c,no_run
+/// struct ExternError {
+///     int32_t code;
+///     char *message; // note: nullable
+/// };
+/// ```
+///
+/// In rust, there are two fields, in this order: `code: ErrorCode`, and `message: *mut c_char`.
+/// Note that ErrorCode is a `#[repr(transparent)]` wrapper around an `i32`, so the first property
+/// is equivalent to an `i32`.
+///
+/// #### The `code` field.
+///
+/// This is the error code, 0 represents success, all other values represent failure. If the `code`
+/// field is nonzero, there should always be a message, and if it's zero, the message will always be
+/// null.
+///
+/// #### The `message` field.
+///
+/// This is a string containing some amount of additional information about the error. If the `code`
+/// property is nonzero, there should always be an error message. Otherwise, this should will be
+/// null.
+///
+/// This string (when not null) is allocated on the rust heap (using this crate's
+/// [`rust_string_to_c`]), and must be freed on it as well. Critically, if there are multiple rust
+/// packages using being used in the same application, it *must be freed on the same heap that
+/// allocated it*, or you will corrupt both heaps.
+///
+/// Typically, this object is managed on the other side of the FFI (on the "FFI consumer"), which
+/// means you must expose a function to release the resources of `message` which can be done easily
+/// using the [`define_string_destructor!`] macro provided by this crate.
+///
+/// If, for some reason, you need to release the resources directly, you may call
+/// `ExternError::release()`. Note that you probably do not need to do this, and it's
+/// intentional that this is not called automatically by implementing `drop`.
+///
+/// ## Example
+///
+/// ```rust,no_run
+/// use ffi_support::{ExternError, ErrorCode};
+///
+/// #[derive(Debug)]
+/// pub enum MyError {
+///     IllegalFoo(String),
+///     InvalidBar(i64),
+///     // ...
+/// }
+///
+/// // Putting these in a module is obviously optional, but it allows documentation, and helps
+/// // avoid accidental reuse.
+/// pub mod error_codes {
+///     // note: -1 and 0 are reserved by ffi_support
```

Comment:
It is enforced, by ErrorCode (which asserts if you try to construct -1). I don't think we can enforce it beyond that though.

No, we can't change the under the hood memory layout, the programmer implementing the "FFI consumer" needs to know what the layout is.

I guess I'd be willing to use i64 if you think it matters, but it's an error code, and I think if you have more than 2^32-1 different possible errors then your program has a problem.

I could be convinced that -1 should be replaced by something less likely to occur. The considerations here are that we do expect the people implementing the "FFI consumer" lib may want to catch this error, even if just for telemetry.

That said, it would be nice to be able to reserve more than just -1 (for example, if we ever want to automate more of the FFI bindings being able to say 'the passed in string isn't valid utf-8' (without needing to know how the user represents this error) would be useful). I'll think of how we can do this / which things to reserve.

Reply:
I think I'll leave it at -1 for now and rethink in the future. Eventually we should have the unsafe part of the FxA android API in this repo and can change these more easily.